internal/uvm/scsi.go

```diff
@@ -50,6 +55,81 @@ func (sm *SCSIMount) Release(ctx context.Context) error {
 	return nil
 }
 
+// If a SCSI mount is read only then we should simply add it to the uvm config. But if it
+// is a scratch layer (i.e writeable mount) then we should make a copy of it.
+func (sm *SCSIMount) Clone(ctx context.Context, vm *UtilityVM, cd *CloneData) (interface{}, error) {
+	var dstVhdPath string = sm.HostPath
+	var err error
+	var dir string
+	conStr := fmt.Sprintf("%d", sm.Controller)
+	lunStr := fmt.Sprintf("%d", sm.LUN)
+	if !sm.IsReadOnly {
+		// Copy this scsi disk
+		// TODO(ambarve): This can be a SCSI mount that belongs to some container
+		// which is being automatically cloned here as a part of UVM cloning
+		// process. We will receive a request for creation of this container
+		// later on which will specify the storage path for this container.
+		// However, that storage location is not available now so we just use
+		// the storage of the uvm instead. Find a better way for handling this.
+		dir, err = ioutil.TempDir(cd.scratchFolder, fmt.Sprintf("clone-mount-%d-%d", sm.Controller, sm.LUN))
+		if err != nil {
+			return nil, fmt.Errorf("Error while creating directory for scsi mounts of clone vm: %s", err)
+		}
+
+		// copy the VHDX
+		dstVhdPath = filepath.Join(dir, filepath.Base(sm.HostPath))
+		log.G(ctx).Debug(fmt.Sprintf("cloning scsi mount with hostPath: %s, Controller: %d, LUN: %d to: %s", sm.HostPath, sm.Controller, sm.LUN, dstVhdPath))
+
+		if err = copyfile.CopyFile(ctx, sm.HostPath, dstVhdPath, true); err != nil {
+			return nil, err
+		}
+
+		if err = wclayer.GrantVmAccess(ctx, cd.UVMID, dstVhdPath); err != nil {
+			os.Remove(dstVhdPath)
+			return nil, err
+		}
+
+	}
+
+	cd.doc.VirtualMachine.Devices.Scsi[conStr].Attachments[lunStr] = hcsschema.Attachment{
+		Path: dstVhdPath,
+		// TODO(ambarve) Get the correct type here
+		Type_: "VirtualDisk",
+	}
+
+	clonedScsiMount := SCSIMount{
+		vm:         vm,
+		HostPath:   dstVhdPath,
+		UVMPath:    sm.UVMPath,
+		Controller: sm.Controller,
+		LUN:        sm.LUN,
+		refCount:   1,
+		IsReadOnly: sm.IsReadOnly,
+	}
+
+	vm.scsiLocations[sm.Controller][sm.LUN] = &clonedScsiMount
+
+	return &clonedScsiMount, nil
+}
+
+func (sm *SCSIMount) GobEncode() ([]byte, error) {
```

Comment:
Is this a placeholder until it's actually encoded with gob? I might be missing something so fill me in.

Reply:
This was meant to be referencing how it's JSON marshaled right now in the line below. Accidentally selected the function name.

Reply:
The same question follows suit for the other GobEncode funcs as well

Reply:
Default GobEncoder will encode every member that is part of this struct. I was having some issues with that earlier and that's why I had added these methods. But I think we can remove them. I will fix this.